File: src/main/kotlin/com/ywc/spark/kafka/serialization/AvroDeserializer.java
package com.ywc.spark.kafka.serialization;

import com.ywc.spark.kafka.Topic;
import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.avro.specific.SpecificRecordBase;
import org.apache.kafka.common.serialization.Deserializer;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Map;

/**
 * @author yanweichen
 * @date 2018/6/19
 */
public class AvroDeserializer implements Deserializer<SpecificRecordBase> {
    @Override
    public void configure(Map<String, ?> configs, boolean isKey) {

    }

    @Override
    public SpecificRecordBase deserialize(String topic, byte[] data) {
        SpecificDatumReader<SpecificRecordBase> userDatumReader =
                new SpecificDatumReader<>(Topic.Companion.matchFor(topic).getTopicType().getSchema());
        BinaryDecoder binaryDecoder = DecoderFactory.get()
                .directBinaryDecoder(new ByteArrayInputStream(data), null);
        try {
            return userDatumReader.read(null, binaryDecoder);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public void close() {

    }
}

File: src/main/kotlin/com/ywc/spark/kafka/serialization/AvroDeserializer.kt
package com.ywc.spark.kafka.serialization

import org.apache.avro.specific.SpecificRecordBase
import org.apache.kafka.common.serialization.Deserializer
import com.ywc.spark.kafka.Topic
import java.io.ByteArrayInputStream
import org.apache.avro.io.DecoderFactory
import org.apache.avro.specific.SpecificDatumReader
import com.sun.xml.internal.ws.encoding.soap.DeserializationException
import java.io.IOException


class AvroDeserializer<T : SpecificRecordBase> : Deserializer<T> {

    override fun configure(configs: MutableMap<String, *>?, isKey: Boolean) {

    }

    override fun deserialize(topic: String, data: ByteArray?): T {
        val userDatumReader = SpecificDatumReader<T>(Topic.matchFor(topic).topicType.schema)
        val binaryEncoder = DecoderFactory.get()
                .directBinaryDecoder(ByteArrayInputStream(data), null)
        try {
            return userDatumReader.read(null, binaryEncoder)
        } catch (e: IOException) {
            throw DeserializationException(e.message)
        }

    }

    override fun close() {

    }
}